pep-9999.rst

```diff
@@ -0,0 +1,669 @@
+			a = a + 1
+			return a
+		else:
+			x = x + 1
+			return x
+
+	freeze(foo)
+	foo(41)  # OK, returns 42
+	foo()    # Throws NotWriteableError
+
+This example illustrates two things. The first call to ``foo(41)``
+shows that local variables on the frame of a frozen function are
+mutable. The second call shows that captured variables are not.
+Note that the default value of ``a`` will be frozen when ``foo``
+is frozen. Thus, the problem of side-effects on default values
+on parameters is avoided.
+
+Frozen function objects that access globals, e.g. through a call
+to ``globals()`` will throw an exception when called.
+
+
+Implementation Details
+======================
+
+1. Introduce an ``is_immutable`` property on objects that tracks whether or
+   not they are frozen. The status is accessible through ``_Py_ISIMMUTABLE``
+   in the C API and in Python code through the ``isimmutable`` builtin.
+2. Modify object mutation operations (``PyObject_SetAttr``,
+   ``PyDict_SetItem``, ``PyList_SetItem``, etc.) to check the 
+   flag and raise an error when appropriate.
+3. Implement ``freeze(obj)``, ensuring it traverses object references
+   safely, including cycle detection.
+4. The builtin ``freezeglobals()`` freezes the globals dictionary and
+   builtins. This is needed for multi-threaded scenarios such as
+   `PEP 703`_ and our planned extensions.
+
+
+Backward Compatibility
+======================
+
+This proposal is fully backward-compatible, as no existing Python
+code will be affected unless it explicitly calls ``freeze(obj)``.
+Frozen objects will raise errors only when mutation is attempted.
+
+
+Performance Implications
+========================
+
+The cost of checking for immutability violations is
+an extra dereference of checking the flag on writes.
+There are implementation-specific issues, such as
+various changes based on how and where the bit is stolen.
+
+
+Alternatives Considered
+=======================
+
+1. Shallow Freezing: Only mark the top-level object as immutable.
+   This would be less effective for ensuring true immutability
+   across references. In particular, this would not make it safe
+   to share the results of ``freeze()`` across threads without risking
+   data-race errors. Shallow immutability is not strong enough to support 
+   sharing immutable objects across subinterpreters (see extensions).
+2. Copy-on-Write Immutability: Instead of raising errors on
+   mutation, create a modified copy. However, this changes object
+   identity semantics and is less predictable. Support for copy-on-write
+   may be added later, if a suitable design can be found.
+3. Immutable Subclasses: Introduce ImmutableDict, ImmutableList,
+   etc., instead of freezing existing objects. However, this does
+   not generalize well to arbitrary objects and adds considerable
+   complexity to all code bases.
+4. Deep freezing immutable copies as proposed in `PEP 351: The 
+   freeze protocol <https://peps.python.org/pep-0351/>`_. That PEP
```

Comment:
Getting Barry's (author of PEP 351) advance feedback on this PEP might be helpful.